### PR TITLE
Project MongoDB by Nasim Mahzoun

### PR DESCRIPTION
**Pull request by Nasimmhn**
endpoints example: 
https://mongo-api-by-nasim.herokuapp.com/
https://mongo-api-by-nasim.herokuapp.com/shows
https://mongo-api-by-nasim.herokuapp.com/shows/id/81082007/
https://mongo-api-by-nasim.herokuapp.com/shows?title=people
https://mongo-api-by-nasim.herokuapp.com/shows?year=2015
https://mongo-api-by-nasim.herokuapp.com/shows?listed_in=stand-up
https://mongo-api-by-nasim.herokuapp.com/shows?title=strong&year=2017&linsted_in=dramas

I did also the frontend part =>  https://react-movie-search-by-nasim.netlify.com/
I did something like Movie-Project, using React route. I didnt have time for styling but I proudly can say, this is my very first and tiny fullstack project 🤓🚩

### CODE REVIEW

File: server.js [hunk 1]
```diff
@@ -2,23 +2,73 @@ import express from 'express'
...
-// import goldenGlobesData from './data/golden-globes.json'
-// import avocadoSalesData from './data/avocado-sales.json'
-// import booksData from './data/books.json'
-// import netflixData from './data/netflix-titles.json'
-// import topMusicData from './data/top-music.json'
 
-const mongoUrl = process.env.MONGO_URL || "mongodb://localhost/project-mongo"
+
+//setup connection to mongodb
+const mongoUrl = process.env.MONGO_URL || "mongodb://localhost/shows"
```
Comment (puj) [inline server.js]: Good naming!

File: server.js [hunk 2]
```diff
@@ -2,23 +2,73 @@ import express from 'express'
...
+  listed_in: {
+    type: String
+  },
+  description: {
+    type: String
+  },
+  type: {
+    type: String
+  },
+})
```
Comment (puj) [inline server.js]: Could the `unique:true` property be applied on one of these?

File: server.js [hunk 3]
```diff
@@ -32,6 +82,62 @@ app.get('/', (req, res) => {
...
+app.get("/shows", async (req, res) => {
+  let queryObj = {}
+
+  let startIndex, perPage
+  if (req.query.perPage) {
+    perPage = +req.query.perPage
+  }
+  if (req.query.page && req.query.perPage) {
+    startIndex = perPage * (+req.query.page - 1)
+  }
```
Comment (puj) [inline server.js]: Excellent naming on your paging variables 💯

File: server.js [hunk 4]
```diff
@@ -32,6 +82,62 @@ app.get('/', (req, res) => {
...
+  if (req.query.listed_in) { queryObj['listed_in'] = new RegExp(req.query.listed_in, 'i') }
+  if (req.query.year) { queryObj['release_year'] = req.query.year }
+  if (req.query.title) { queryObj['title'] = new RegExp(req.query.title, 'i') }
+
+  Show.find(queryObj).sort('title')
```
Comment (puj) [inline server.js]: Nice choice of default sort field

File: server.js [hunk 5]
```diff
@@ -32,6 +82,62 @@ app.get('/', (req, res) => {
...
+    .then((results) => {
+      let resultsObj = {
+        "total_shows": results.length
+      }
```
Comment (puj) [inline server.js]: Yes yes yes yes yes! Fantastic!  The client needs this information!

File: server.js [hunk 6]
```diff
@@ -32,6 +82,62 @@ app.get('/', (req, res) => {
...
+      // Successful
+      if (req.query.page && req.query.perPage) {
+        resultsObj.shows = results.slice(startIndex, startIndex + perPage)
```
Comment (puj) [inline server.js]: Why not use Mongoose's `.limit()` and `.skip()`?

File: server.js [hunk 7]
```diff
@@ -32,6 +82,62 @@ app.get('/', (req, res) => {
...
+        res.json(resultsObj)
+      } else {
+
+        resultsObj.shows = results
+        res.json(resultsObj)
+      }
```
Comment (puj) [inline server.js]: Is it possible to rewrite this without the `else`?

File: server.js [hunk 8]
```diff
@@ -32,6 +82,62 @@ app.get('/', (req, res) => {
...
+    }).catch((err) => {
+      //Error - Failure
+      res.json({ message: 'Cannot find this show', err: err })
+    })
+})
+
+
+// path-params to be able to find a specefic show //
+// http://localhost:8080/shows/id/81082007/
+app.get("/shows/id/:id", async (req, res) => {
```
Comment (puj) [inline server.js]: You could probably use `/shows/:id` common convention in REST

**Comment by Nasimmhn** [comment]
> Fantastic job pushing the stretch goals! Great understanding of paging so far 😄
> 
> Where is the frontend code?!? I didn't know if the frontend actually used the backend or if you are saving that for later 😉
> 
> Well done! Seems like you enjoyed this one

> This is the right deployed link for the frontend part => https://react-movie-search-by-nasim.netlify.com/
>Apparently I forgot to change the API-URL from localhost to Heroku :)

File: server.js (hunk 1 above)
Comment (Nasimmhn) [reply server.js]: @puj This is the right deployed link for the frontend part => https://react-movie-search-by-nasim.netlify.com/
Apparently I forgot to change the API-URL from localhost to Heroku :)

**Pull request closed without merging**